### PR TITLE
Update admin on webhook

### PR DESCRIPTION
**Pull request by yoshwata**
## Context

Our cluster has many request by no admin exists recently. 

Admin add/delete function is implemented on start or sync from UI.  Webhook is also a function that user frequently use, so I'd like to port it on webhook.

## Objective

Users are registered as pipeline's admin when they trigger from github push or open PR.

## References

https://github.com/screwdriver-cd/screwdriver/pull/1106

### CODE REVIEW

File: plugins/webhooks/index.js [hunk 1]
```diff
@@ -2,10 +2,76 @@
...
+ * @param  {UserFactory}    userFactory     UserFactory object
+ * @param  {String}         username        Username of user
+ * @param  {String}         scmContext      Scm which pipeline's repository exists in
+ * @param  {Pipeline}       pipeline        Pipeline object
+ */
+async function checkAndUpdateAdmins(userFactory, username, scmContext, pipeline) {
+    try {
+        await userFactory.get({ username, scmContext })
+            .then((user) => {
+                if (!pipeline) {
```
Comment (tk3fftk) [inline plugins/webhooks/index.js]: Is this check needed?

File: plugins/webhooks/index.js [hunk 2]
```diff
@@ -2,10 +2,76 @@
...
+/**
+ * Update admins array after check existance of pipeline, user and permission
+ * @param  {UserFactory}    userFactory     UserFactory object
+ * @param  {String}         username        Username of user
+ * @param  {String}         scmContext      Scm which pipeline's repository exists in
+ * @param  {Pipeline}       pipeline        Pipeline object
+ */
+async function checkAndUpdateAdmins(userFactory, username, scmContext, pipeline) {
+    try {
+        await userFactory.get({ username, scmContext })
```
Comment (tk3fftk) [inline plugins/webhooks/index.js]: I think the above example is more simple when we use `await`.
```
try{
  const user = await userFactory.get({ username, scmContext });
  const userPermissions = await user.getPermissions(pipeline.scmUri);

  return updateAdmins(
                userPermissions,
                pipeline,
                username));

} catch (err) {...}
```

File: test/plugins/webhooks.test.js [hunk 3]
```diff
@@ -502,15 +509,75 @@ describe('github plugin test', () => {
                 });
             });
 
-            it('returns 500 when failed', () => {
-                eventFactoryMock.create.rejects(new Error('Failed to start'));
+            it('reuturns 200 when user is registered as admin', () => {
```
[the remark this answers is not in the document]
Comment (yoshwata) [reply test/plugins/webhooks.test.js]: I fixed, thank you.

File: plugins/webhooks/index.js (hunk 1 above)
Comment (yoshwata) [reply plugins/webhooks/index.js]: No need to do that, I fixed. Thank you.

File: plugins/webhooks/index.js (hunk 2 above)
Comment (yoshwata) [reply plugins/webhooks/index.js]: It's nice idea. I fixed. Thank you.

File: plugins/webhooks/index.js [hunk 4]
```diff
@@ -2,10 +2,48 @@
...
+    try {
+        const user = await userFactory.get({ username, scmContext });
+        const userPermissions = await user.getPermissions(pipeline.scmUri);
+        const newAdmins = pipeline.admins;
+
+        // Delete user from admin list if bad permissions
+        if (!userPermissions.push) {
+            delete newAdmins[username];
+            // This is needed to make admins dirty and update db
+            pipeline.admins = newAdmins;
```
Comment (tk3fftk) [inline plugins/webhooks/index.js]: `pipeline.admins = newAdmins;` and `return pipeline.update();` are duplicated with the below `if` statement. We can aggregate these.

File: plugins/webhooks/index.js [hunk 5]
```diff
@@ -421,14 +459,14 @@ function pullRequestEvent(pluginOptions, request, reply, parsed) {
                         restrictPR
                     };
 
+                    updateAdmins(userFactory, username, scmContext, pipeline);
```
Comment (wahapo) [inline plugins/webhooks/index.js]: We do not need await here? I am worried about when `updateAdmins` will be completed.

Comment (yoshwata) [reply plugins/webhooks/index.js]: You are right. I add `await`. Thank you.

**Pull request merged**